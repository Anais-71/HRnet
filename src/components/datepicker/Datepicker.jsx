import React, { useState } from 'react'
import './Datepicker.css'

const Datepicker = ({ idPrefix }) => {
  const [isOpen, setIsOpen] = useState(false)
  const [selectedDate, setSelectedDate] = useState('')
  const [currentMonth, setCurrentMonth] = useState(new Date().getMonth())
  const [currentYear, setCurrentYear] = useState(new Date().getFullYear())

  const months = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ]

  const weekdays = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']

  const daysInMonth = (month, year) => new Date(year, month + 1, 0).getDate()

  const handleDayClick = (day) => {
    const formattedDate = `${day}/${currentMonth + 1}/${currentYear}`
    setSelectedDate(formattedDate)
    setIsOpen(false)
  }

  const renderCalendar = () => {
    const days = []
    const totalDays = daysInMonth(currentMonth, currentYear)
    const firstDayIndex = new Date(currentYear, currentMonth, 1).getDay()

    // Render empty slots before first day
    for (let i = 0; i < firstDayIndex; i++) {
      days.push(
        <div key={`empty-${i}`} className="datepicker__grid--day empty"></div>,
      )
    }

    // Render days of the month
    for (let day = 1; day <= totalDays; day++) {
      days.push(
        <div
          key={day}
          className="datepicker__grid--day"
          onClick={() => handleDayClick(day)}
        >
          {day}
        </div>,
      )
    }

    return (
      <div className="datepicker__grid">
        {weekdays.map((weekday, index) => (
          <div key={index} className="datepicker__grid--weekday">
            {weekday}
          </div>
        ))}
        {days}
      </div>
    )
  }

  return (
    <div className="datepicker">
      <label htmlFor={idPrefix}>Select a date:</label>{' '}
      <input
        type="text"
        id={idPrefix}
        value={selectedDate}
        readOnly
        onClick={() => setIsOpen(!isOpen)}
        placeholder="Please select a date"
        aria-haspopup="true"
        aria-expanded={isOpen}
        autoComplete="off"
      />
      {isOpen && (
        <div className="datepicker__calendar" role="dialog" aria-modal="true">
          <div className="datepicker__controls">
            <label htmlFor={`${idPrefix}-month`}>Month:</label>{' '}
            <select
              id={`${idPrefix}-month`}
              value={currentMonth}
              onChange={(e) => setCurrentMonth(Number(e.target.value))}
            >
              {months.map((month, index) => (
                <option key={index} value={index}>
                  {month}
                </option>
              ))}
            </select>
            <label htmlFor={`${idPrefix}-year`}>Year:</label>{' '}
            <select
              id={`${idPrefix}-year`}
              value={currentYear}
              onChange={(e) => setCurrentYear(Number(e.target.value))}
            >
              {Array.from({ length: 76 }, (_, index) => 1950 + index).map(
                (year) => (
                  <option key={year} value={year}>
                    {year}
                  </option>
                ),
              )}
            </select>
          </div>
          {renderCalendar()}
        </div>
      )}
    </div>
  )
}

export default Datepicker
