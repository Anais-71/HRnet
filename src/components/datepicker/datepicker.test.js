import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import Datepicker from './Datepicker' // Ensure the import path is correct

/**
 * Test suite for the Datepicker component.
 */
describe('Given I use the Datepicker plugin', () => {
  /**
   * Renders the Datepicker component before each test.
   */
  beforeEach(() => {
    render(<Datepicker idPrefix="datepicker" />)
  })

  /**
   * Test to check if the input field is rendered.
   */
  test('Then the input field is rendered', () => {
    const input = screen.getByPlaceholderText(/please select a date/i)
    expect(input).toBeInTheDocument()
  })

  /**
   * Test to verify that the calendar opens upon input click.
   */
  test('Then the calendar opens on input click', () => {
    const input = screen.getByPlaceholderText(/please select a date/i)
    fireEvent.click(input)
    const calendar = screen.getByRole('dialog')
    expect(calendar).toBeInTheDocument()
  })

  /**
   * Test to check if the selected date is displayed in the input field.
   */
  test('When I select a date from the calendar, Then the date is displayed in the input', () => {
    const input = screen.getByPlaceholderText(/please select a date/i)
    fireEvent.click(input)

    const dayToSelect = screen.getByText('1')
    fireEvent.click(dayToSelect)
    expect(input.value).toBe('1/10/2024')
  })

  /**
   * Test to verify the correct month is rendered when changed using the month selector.
   */
  test('When I change the month using month selector, Then the correct month is rendered', () => {
    const input = screen.getByPlaceholderText(/please select a date/i)
    fireEvent.click(input)

    const monthSelect = screen.getByLabelText(/month/i)
    expect(monthSelect).toBeInTheDocument()

    fireEvent.change(monthSelect, { target: { value: '2' } }) // Change to March

    const newMonthDay = screen.getByText('1')
    expect(newMonthDay).toBeInTheDocument()
  })

  /**
   * Test to verify the correct year is rendered when changed using the year selector.
   */
  test('When I change the year using year selector, Then the correct year should be rendered', () => {
    const input = screen.getByPlaceholderText(/please select a date/i)
    fireEvent.click(input)

    const yearSelect = screen.getByLabelText(/year/i)
    expect(yearSelect).toBeInTheDocument()

    fireEvent.change(yearSelect, { target: { value: '2025' } })

    const yearDisplay = screen.getByText('1')
    expect(yearDisplay).toBeInTheDocument()
  })

  /**
   * Test to check if the calendar closes when clicking on the selected date.
   */
  test('Given I selected a date, When I click on the selected date, Then the calendar is closed', () => {
    const input = screen.getByPlaceholderText(/please select a date/i)
    fireEvent.click(input)

    const dayToSelect = screen.getByText('1')
    fireEvent.click(dayToSelect)
    expect(input.value).toBe('1/10/2024')

    fireEvent.click(dayToSelect)
    const calendar = screen.queryByRole('dialog')
    expect(calendar).not.toBeInTheDocument()
  })
})

/**
 * Snapshot test for the Datepicker component.
 */
test('Datepicker component matches snapshot', () => {
  const { asFragment } = render(<Datepicker idPrefix="datepicker" />)
  expect(asFragment()).toMatchSnapshot()
})
