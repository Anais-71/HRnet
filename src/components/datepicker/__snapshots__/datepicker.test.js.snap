// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datepicker component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="datepicker"
  >
    <input
      aria-expanded="false"
      aria-haspopup="true"
      autocomplete="off"
      id="datepicker"
      placeholder="Please select a date"
      readonly=""
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
